File: cometd-java/bayeux-api/src/main/java/org/cometd/bayeux/Transport.java
/*
 * Copyright (c) 2008-2020 the original author or authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.cometd.bayeux;

import java.util.Set;

/**
 * <p>A transport abstract the details of the protocol used to send
 * Bayeux messages over the network, for example using HTTP or using
 * WebSocket.</p>
 * <p>{@link Transport}s have well known names and both a Bayeux client
 * and a Bayeux server can negotiate the transport they want to use by
 * exchanging the list of supported transport names.</p>
 * <p>Transports can be configured using <em>options</em>. The transport
 * implementation provides a set of {@link #getOptionNames() option names} that
 * it uses to configure itself and an {@link #getOptionPrefix() option prefix}
 * that allows specific tuning of the configuration.
 * Option prefixes may be composed of segments separated by the "." character.</p>
 * <p>For example, imagine to configure the transports for normal long polling,
 * for JSONP long polling and for WebSocket. All provide a common option name
 * called "timeout" and the JSONP long polling transport provides also a specific
 * option name called "callback".
 * The normal long polling transport has prefix "long-polling.json",
 * the JSONP long polling transport has prefix "long-polling.jsonp" and the
 * WebSocket long polling transport has prefix "ws". The first two prefixes
 * have 2 segments.</p>
 * <p>The configurator will asks the transports the set of option names, obtaining
 * ["timeout", "callback"]; then will ask each transport its prefix, obtaining
 * ["long-polling.json", "long-polling.jsonp"].
 * The configurator can now look in the configuration (for example a properties
 * file or servlet init parameters) for entries that combine the option names and
 * option prefix segments, such as:</p>
 * <ul>
 * <li>"timeout": specifies the timeout for all transports</li>
 * <li>"long-polling.timeout": specifies the timeout for both normal long polling
 * transport and JSONP long polling transport, but not for the WebSocket transport</li>
 * <li>"long-polling.jsonp.timeout": specifies the timeout for JSONP long polling
 * transport overriding more generic entries</li>
 * <li>"ws.timeout": specifies the timeout for WebSocket transport overriding more
 * generic entries</li>
 * <li>"long-polling.jsonp.callback": specifies the "callback" parameter for the
 * JSONP long polling transport.</li>
 * </ul>
 */
public interface Transport {
    /**
     * @return The well known name of this transport, used in transport negotiations
     * @see Bayeux#getAllowedTransports()
     */
    String getName();

    /**
     * @param name the configuration option name
     * @return the configuration option with the given {@code qualifiedName}
     * @see #getOptionNames()
     */
    Object getOption(String name);

    /**
     * @return the set of configuration options
     * @see #getOption(String)
     */
    Set<String> getOptionNames();

    /**
     * Specifies an option prefix made of string segments separated by the "."
     * character, used to override more generic configuration entries.
     *
     * @return the option prefix for this transport.
     */
    String getOptionPrefix();
}

File: cometd-javascript/cometd-javascript-common/src/main/webapp/js/cometd/Transport.js
/*
 * Copyright (c) 2008 the original author or authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * Base class with the common functionality for transports.
 */
export class Transport {
    #type;
    #cometd;
    #url;

    /**
     * Function invoked just after a transport has been successfully registered.
     * @param type the type of transport (for example "long-polling")
     * @param cometd the cometd object this transport has been registered to
     * @see #unregistered()
     */
    registered(type, cometd) {
        this.#type = type;
        this.#cometd = cometd;
    };

    /**
     * Function invoked just after a transport has been successfully unregistered.
     * @see #registered(type, cometd)
     */
    unregistered() {
        this.#type = null;
        this.#cometd = null;
    };

    get cometd() {
        return this.#cometd;
    }

    notifyTransportTimeout(messages) {
        const callbacks = this.cometd._getTransportListeners("timeout");
        if (callbacks) {
            for (let i = 0; i < callbacks.length; ++i) {
                const listener = callbacks[i];
                try {
                    const result = listener.call(this, messages);
                    if (typeof result === "number" && result > 0) {
                        return result;
                    }
                } catch (x) {
                    this.cometd._info("Exception during execution of transport listener", listener, x);
                }
            }
        }
        return 0;
    };

    debug() {
        this.cometd._debug.apply(this.cometd, arguments);
    };

    get configuration() {
        return this.cometd.getConfiguration();
    }

    get advice() {
        return this.cometd.getAdvice();
    }

    setTimeout(funktion, delay) {
        return this.cometd.setTimeout(funktion, delay);
    }

    clearTimeout(id) {
        this.cometd.clearTimeout(id);
    };

    convertToJSON(messages) {
        const maxSize = this.configuration.maxSendBayeuxMessageSize;
        let result = "[";
        for (let i = 0; i < messages.length; ++i) {
            if (i > 0) {
                result += ",";
            }
            const message = messages[i];
            const json = JSON.stringify(message);
            if (json.length > maxSize) {
                throw new Error("maxSendBayeuxMessageSize " + maxSize + " exceeded");
            }
            result += json;
        }
        result += "]";
        return result;
    };

    /**
     * Converts the given response into an array of bayeux messages
     * @param response the response to convert
     * @return an array of bayeux messages obtained by converting the response
     */
    convertToMessages(response) {
        if (response === undefined || response === null) {
            return [];
        }
        if (typeof response === "string" || response instanceof String) {
            try {
                return JSON.parse(response);
            } catch (x) {
                this.debug("Could not convert to JSON the following string:", response);
                throw x;
            }
        }
        if (Array.isArray(response)) {
            return response;
        }
        if (response instanceof Object) {
            return [response];
        }
        throw new Error("Conversion Error " + response + ", typeof " + typeof response);
    };

    /**
     * Returns whether this transport can work for the given version and cross domain communication case.
     * @param version a string indicating the transport version
     * @param crossDomain a boolean indicating whether the communication is cross domain
     * @param url the URL to connect to
     * @return true if this transport can work for the given version and cross domain communication case,
     * false otherwise
     */
    accept(version, crossDomain, url) {
        throw new Error("Abstract");
    };

    /**
     * Returns the type of this transport.
     * @see #registered(type, cometd)
     */
    get type() {
        return this.#type;
    }

    get url() {
        return this.#url;
    }

    set url(url) {
        this.#url = url;
    };

    send(envelope, metaConnect) {
        throw new Error("Abstract");
    };

    reset(init) {
        this.debug("Transport", this.type, "reset", init ? "initial" : "retry");
    };

    abort() {
        this.debug("Transport", this.type, "aborted");
    };

    toString() {
        return this.type;
    };
}
